### PR TITLE
[POC] feat: add support for session refresh for safari

### PR DESCRIPTION
**Pull request by hunger-programmer**
ref: UUI-683

### CODE REVIEW

**Review by bogdan-niculescu-sch — APPROVED** [review]
looks good to me

File: __tests__/identity.js [hunk 1]
```diff
@@ -484,6 +484,21 @@ describe('Identity', () => {
...
+                identity._sessionService.fetch.mockReturnValueOnce(
+                    ({ ok: true, json: () => Fixtures.sessionNeedsToBeRefreshedResponse })
+                )
+                const spy = jest.spyOn(identity, 'emit');
+
+                await identity.hasSession();
+
+                expect(spy).toHaveProperty('mock.calls.length');
+                expect(spy.mock.calls.length).toBeGreaterThan(0);
+                expect(spy.mock.calls.some(c => c[0] === 'redirectToSessionService')).toBe(true);
```
Comment (hunger-programmer) [inline __tests__/identity.js:499]: This test can be simplified
```                expect(spy).toHaveBeenCalledTimes(1);
                expect(spy).toHaveBeenCalledWith('redirectToSessionService');```

File: src/identity.js [hunk 2]
```diff
@@ -508,10 +516,23 @@ export class Identity extends EventEmitter {
...
-            if (sessionData && this._enableSessionCaching) {
-                const expiresIn = 1000 * (sessionData.expiresIn || 300);
-                this.cache.set(HAS_SESSION_CACHE_KEY, sessionData, expiresIn);
+            if(sessionData){
+                // for expiring session and safari browser do full page redirect to gain new session
+                if(_checkRedirectionNeed(sessionData)){
+                    const client_sdrn = `sdrn:${NAMESPACE[this.env]}:client:${this.clientId}`;
+                    const redirectBackUrl = this.redirectUri.substring(0 , this.redirectUri.lastIndexOf('/'));
+                    const params = { redirect_uri: redirectBackUrl, client_sdrn:  client_sdrn};
+                    this.emit('redirectToSessionService');
```
[the remark this answers is not in the document]
Comment (michal-oleniacz-stp) [reply src/identity.js:525]: Wouldn't this event allow the brands to save any progress on critical flows to i.e. localStorage before the redirect happens?

**Review by bogdan-niculescu-sch — CHANGES_REQUESTED** [review]
changing to request changes to not merge by mistake since this needs to be a feature😁

File: .github/workflows/npm-publish.yml [hunk 3]
```diff
@@ -38,6 +38,6 @@ jobs:
             -   name: Build application
                 run: npm run build
             -   name: Publish to npm
-                run: npm publish --access=public
+                run: npm publish --access=public --tag beta #TODO: To be removed, for PoC purposes only
```
Comment (MichalOleniacz) [inline .github/workflows/npm-publish.yml:41]: This should be removed

Comment (michal-oleniacz-stp) [inline .github/workflows/npm-publish.yml:41]: Beta tag should be removed for final release

File: src/identity.js [hunk 4]
```diff
@@ -511,16 +537,36 @@ export class Identity extends EventEmitter {
...
             }
 
-            if (sessionData && this._enableSessionCaching) {
-                const expiresIn = 1000 * (sessionData.expiresIn || 300);
-                this.cache.set(HAS_SESSION_CACHE_KEY, sessionData, expiresIn);
+            if (sessionData){
+                // for expiring session and safari browser do full page redirect to gain new session
+                if(_checkRedirectionNeed(sessionData)){
+                    if (this._enableSessionCaching) {
+                        const expiresIn = 1000 * (sessionData.expiresIn || 300);
```
Comment (michal-oleniacz-stp) [inline src/identity.js:544]: In this case, `sessionData` is an object with one key "redirectURL".
We should consider creating a separate field for this.

File: src/identity.js [hunk 5]
```diff
@@ -511,16 +537,36 @@ export class Identity extends EventEmitter {
                 throw err;
             }
 
-            if (sessionData && this._enableSessionCaching) {
-                const expiresIn = 1000 * (sessionData.expiresIn || 300);
-                this.cache.set(HAS_SESSION_CACHE_KEY, sessionData, expiresIn);
+            if (sessionData){
+                // for expiring session and safari browser do full page redirect to gain new session
+                if(_checkRedirectionNeed(sessionData)){
+                    if (this._enableSessionCaching) {
```
Comment (michal-oleniacz-stp) [inline src/identity.js:543]: Can this "semaphore" logic be extracted into separate methods?
`hasSession` does a lot already.

File: src/identity.js [hunk 6]
```diff
@@ -480,9 +491,16 @@ export class Identity extends EventEmitter {
      * @return {Promise<HasSessionSuccessResponse|HasSessionFailureResponse>}
      */
     hasSession() {
+        const checkIfSessionOngoing = this.cache.get("sessionFlowOngoing");
+        if (checkIfSessionOngoing)
```
Comment (michal-oleniacz-stp) [inline src/identity.js:495]: Please remove newline before '{'

**Pull request merged**